### ISSUE TITLE
Crash in AsyncOperation

### ISSUE THREAD

**Issue by matheda** [algolia/algoliasearch-client-swift#592]
We are seeing the following crash in the Algolia SDK.  It is not super frequent, only 117 times in the last 90 days.

![screen shot 2018-11-10 at 8 45 26 am](https://user-images.githubusercontent.com/1895008/48302067-5d4fc800-e4c5-11e8-8b98-df0000576783.png)

It looks like it is related to the didChangeValue KVM notification on the finished property, so I am guessing whatever is monitoring that property is having an issue.

Thanks!

**Comment by spinach** [comment]
Hey @matheda,

Thanks for reporting this! We're investigating to see the cause of this, and will get back to you!

**Comment by matheda** [comment]
Just wanted to follow up on this, we are still seeing a large number of crashes around it.  In case it is helpful, I ran into someone who found that there is a KVO issue in Swift:  https://bugs.swift.org/browse/SR-6795

**Comment by spinach** [comment]
Hey @matheda, 

Thanks for linking to that issue, that gives us something to try on our end to try to reproduce the error. 

I also want to assure you that we are not ignoring this issue and actively tried to reproduce the issue on our end in order to approve the fixes that were proposed. Unfortunately, no one was able to reproduce the crash in a UT, nor gave us the concrete steps to do so, and that's why it wouldn't make sense for us to push a change without knowing if it would address an actual issue.

Also, in the original post, you had mentioned crashes of 117 times in 90 days, around 1.3 crashes per day for all your users. I'm curious if this number increased?

Thanks!

**Comment by matheda** [comment]
In the last 30 days we have seen it 106 times for 104 unique users.

**Comment by spinach** [comment]
@matheda Thanks for sharing that. We tried running multiple tests in order to reproduce the KVO issue, we tried them using both Swift 4 and Swift 5 since the bug has been marked as resolved, but to no avail. We tried running 10 million async operations in different conditions to try to reproduce but we can't seem to. 

We were also thinking that maybe this crash is happening while the app is in the background but that is pretty hard to reproduce. We'll do more experiments on our side to try to reproduce it. 

A small suggestion I can also give you, but is risky: You can try to apply the fix of suggested PR https://github.com/algolia/algoliasearch-client-swift/pull/603 and track that on your end to see if the crash goes away. From the description of the PR, it seems that they are happy with the solution, and to make sure, I just asked them to see if they have concrete results proving that the crash got solved.

**Comment by matheda** [comment]
DO NOT MERGE THIS PR... more details later today

**Comment by matheda** [comment]
Last night we released an updated version of the AllTrails app which contains the change from PR #603.  We had the change in place during our internal testing and didn't run into any issues.  However, very soon after release we started getting velocity errors on two new crashes in the Algolia SDK.  In less than 24 hours we have seen ~1300 crashes between the two.  I am attaching the stack traces for the two crashes below.  We have reverted the change and are resubmitting a point release today.

So far we have not seen the crash in AsyncOperation that we originally reported, but the occurrence rate of that is so low that it is premature to say if it was addressed.

![Screen Shot 2019-06-13 at 2 58 49 PM](https://user-images.githubusercontent.com/1895008/59460399-ff81b680-8dec-11e9-8c78-877aa6b55708.png)

![Screen Shot 2019-06-13 at 3 00 11 PM](https://user-images.githubusercontent.com/1895008/59460409-05779780-8ded-11e9-9e61-70e673fe0ded.png)

![Screen Shot 2019-06-13 at 3 00 55 PM](https://user-images.githubusercontent.com/1895008/59460419-0c9ea580-8ded-11e9-8b80-337346a76b53.png)
